terraform/install.sh

```diff
@@ -37,25 +39,93 @@ getBillingAccount() {
     log "To list active billing accounts, run:"
     log "gcloud beta billing accounts list --filter open=true"
     exit 1;
-  elif [[ $(echo "${found_accounts}" | wc -l) -gt 1 ]]; then
+  fi
+
+  # store (name:id) info in a map
+  IFS_bak=$IFS
+  declare -A map
+  acc_ids=$(gcloud beta billing accounts list --format="value(displayName,name)" --filter open=true)
+  IFS=$'\n'
+  acc_ids=($acc_ids)
+  for acc in ${acc_ids[@]}
+  do
+    IFS=$'\t'
+    acc=($acc)
+    IFS=$'\n'
+    map[${acc[0]}]=${acc[1]}
+  done
+
+  if [[ $(echo "${found_accounts}" | wc -l) -gt 1 ]]; then
       log "Which billing account would you like to use?:"
-      IFS_bak=$IFS
       IFS=$'\n'
       select opt in ${found_accounts} "cancel"; do
         if [[ "${opt}" == "cancel" ]]; then
           exit 0
         elif [[ -z "${opt}" ]]; then
           log "invalid response"
         else
-          billing_acct=$opt
+          billing_acct=${opt}
+          billing_id=${map[$billing_acct]}
           break
         fi
       done
-      IFS=$IFS_bak
   else
     billing_acct=${found_accounts}
+    billing_id=${map[$found_accounts]}
+  fi
+  IFS=$IFS_bak
+}
+
+getProject() {
+  log "Checking for project list"
+  found_projects=$(gcloud projects list --filter="name:stackdriver-sandbox-*" --format="value(projectId)")
+  if [ -z "$found_projects" ] || [[ ${#found_projects[@]} -eq 0 ]]; then
+    create_flag=true
+  else
+      log "Which project would you like to use?:"
+      IFS_bak=$IFS
+      IFS=$'\n'
+      select opt in ${found_projects} "create a new Sandbox" "cancel"; do
+        if [[ "${opt}" == "cancel" ]]; then
+          exit 0
+        elif [[ "${opt}" == "create a new Sandbox" ]]; then
+          create_flag=true
+          log "create a new Sandbox!"
+          break
+        elif [[ -z "${opt}" ]]; then
+          log "invalid response"
+        else
+          project_id=$opt
+          bucket_name="$project_id-bucket"
+          break
+        fi
+      done
+      IFS=$IFS_bak
+  fi
+}
+
+createProject() {
+  if [ "$create_flag" = true ]; then
+    # generate random id
+    project_id="stackdriver-sandbox-$(od -N 4 -t uL -An /dev/urandom | tr -d " ")"
+    bucket_name="stackdriver-sandbox-state"
+    # create project
+    if [ -z "$folder_id" ]; then
+      gcloud projects create "$project_id" --name="Stackdriver Sandbox Demo"
+    else
+      gcloud projects create "$project_id" --name="Stackdriver Sandbox Demo" --folder="$folder_id"
+    fi;
+    # link billing account
+    gcloud beta billing projects link "$project_id" --billing-account="$billing_id"
+    # confirm billing account linked
+    check_billing=$(gcloud beta billing projects list --billing-account="$billing_id" --format="value(project_id)" --filter="project_id:$project_id")
+    while [ -z "$check_billing" ]; do
+      log "waiting for billing account to be linked"
+      sleep 1
```

Comment:
I'm a bit nervoud about leaving this potentially infinite loop here. Maybe you should add a TRIES variable like we use here: https://github.com/GoogleCloudPlatform/stackdriver-sandbox/blob/dab7ab519a04c874058f3b2dc0d552b345590d6e/terraform/install.sh#L100-L102

Reply:
Good idea!

Reply:
The problem is that the while loop doesn't help waiting the project to be linked.